contracts/external/cw-payroll/src/contract.rs

```diff
@@ -0,0 +1,1462 @@
+
+    if end_time <= block_time {
+        return Err(ContractError::InvalidEndTime {});
+    }
+
+    let duration: u128 = (end_time - start_time).into();
+    let balance_amount = balance.amount();
+    let refund: u128 = balance_amount
+        .checked_rem(duration)
+        .ok_or(GenericError::IntegerOverflow {})
+        .unwrap();
+
+    if balance_amount < duration {
+        return Err(ContractError::AmountLessThanDuration {});
+    }
+
+    let claimed = if balance.is_native() {
+        WrappedBalance::default()
+    } else {
+        WrappedBalance::new_cw20(balance.cw20().unwrap().address.clone(), Uint128::new(0))
+    };
+    let stream = Stream {
+        admin: admin.clone(),
+        recipient: recipient.clone(),
+        balance: balance.clone(),
+        claimed_balance: claimed,
+        start_time,
+        end_time,
+        paused_time: None,
+        paused_duration: None,
+        paused: false,
+        title,
+        description,
+        link_id: None,
+        is_detachable: is_detachable.unwrap_or(true),
+    };
+    let id = add_stream(deps, &stream)?;
+
+    let mut response = Response::new()
+        .add_attribute("method", "create_stream")
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("admin", admin.clone())
+        .add_attribute("recipient", recipient)
+        .add_attribute("start_time", start_time.to_string())
+        .add_attribute("end_time", end_time.to_string());
+
+    if refund > 0 {
+        response = response.add_messages(create_balance_transfer_msg(
+            &balance,
+            Some(Uint128::new(refund)),
+            admin.into(),
+        ));
+    }
+    Ok(response)
+}
+
+pub fn execute_receive(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> ContractResult {
+    deps.api.addr_validate(&info.sender.into_string())?;
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+    match msg {
+        ReceiveMsg::CreateStream {
+            admin,
+            start_time,
+            end_time,
+            recipient,
+            is_detachable,
+        } => execute_create_stream(
+            env,
+            deps,
+            StreamParams {
+                admin: admin.unwrap_or_else(|| wrapped.sender.clone()),
+                recipient,
+                balance: WrappedBalance::from(wrapped),
+                start_time,
+                end_time,
+                title: None,
+                description: None,
+                is_detachable,
+            },
+        ),
+    }
+}
+
+pub fn execute_distribute(env: Env, deps: DepsMut, id: u64) -> ContractResult {
+    let mut stream = STREAMS
+        .may_load(deps.storage, id)?
+        .ok_or(ContractError::StreamNotFound { stream_id: id })?;
+
+    let mut msgs: Vec<CosmosMsg> = vec![];
+    let (available_claims, _) = stream.calc_distribution_rate(env.block.time);
+
+    if !stream.can_distribute_more() || available_claims.u128() == 0 {
```

Comment:
shouldn't `avaliable_claims` being nonzero imply that the stream can distribute more?

Reply:
can_distribute_more will be faster and we don't have to call calc_distribution_rate if balance is 0, but the logic could work I moved !stream.can_distribute_more() check upper, thank you for notice)